### PR TITLE
cargo-miri: create stub .d files

### LINKED ISSUE #1724: "dep-bin-cargo-miri-test" is only created on the second `cargo miri run`
To reproduce, run `rm target -rf && cargo miri run -- "" && cargo miri run -- ""` in the `test-cargo-miri` directory:
```
   Compiling byteorder v1.3.4
   Compiling cargo-miri-test v0.1.0 (/home/r/src/rust/miri/test-cargo-miri)
   Compiling issue_1691 v0.1.0 (/home/r/src/rust/miri/test-cargo-miri/issue-1691)
   Compiling cdylib v0.1.0 (/home/r/src/rust/miri/test-cargo-miri/cdylib)
   Compiling issue_1705 v0.1.0 (/home/r/src/rust/miri/test-cargo-miri/issue-1705)
   Compiling issue_1567 v0.1.0 (/home/r/src/rust/miri/test-cargo-miri/issue-1567)
    Finished dev [unoptimized + debuginfo] target(s) in 0.78s
     Running `/home/r/.cargo/bin/cargo-miri target/x86_64-unknown-linux-gnu/debug/cargo-miri-test ''`
0x01020304
main

   Compiling cargo-miri-test v0.1.0 (/home/r/src/rust/miri/test-cargo-miri)
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `/home/r/.cargo/bin/cargo-miri target/x86_64-unknown-linux-gnu/debug/cargo-miri-test ''`
0x01020304
main

```
Notice how in the second run, it is still doing `Compiling cargo-miri-test`. If we do a third run, it does not do this any more.

Running with `CARGO_LOG=cargo::core::compiler::fingerprint=trace` shows that cargo is missing a file called `dep-bin-cargo-miri-test`, which (as of right now on my platform) it expects in `target/x86_64-unknown-linux-gnu/debug/.fingerprint/cargo-miri-test-44bcc890eaab31dc`. After the first run, that folder contains
```
bin-cargo-miri-test  bin-cargo-miri-test.json  invoked.timestamp
```
Only after the second run does it also contain the expected `dep-bin-cargo-miri-test`.

This is not a terribly big deal since the `Compiling` step for a binary that Miri will run does literally nothing, but it's still strange noise in the log, and it would be good to figure out what is going on.

@ehuss could you give some guidance, in particular about who is supposed to create that file and why it might only happen on the second attempt here?

### PR DESCRIPTION
**Pull request by hyd-dev**
Thanks @RalfJung for the discovery: https://github.com/rust-lang/miri/issues/1724#issuecomment-787115693 :heart:
Fixes #1724.

### CODE REVIEW

File: cargo-miri/bin.rs [hunk 1]
```diff
@@ -609,6 +609,14 @@ fn phase_cargo_rustc(mut args: env::Args) {
     let print = get_arg_flag_value("--print").is_some(); // whether this is cargo passing `--print` to get some infos
 
     let store_json = |info: CrateRunInfo| {
+        // Create a stub .d file to stop Cargo from "rebuilding" the crate:
+        // https://github.com/rust-lang/miri/issues/1724#issuecomment-787115693
+        // As we store a JSON file instead of building the crate here, an empty file is fine.
+        let dep_info_name = out_filename("", ".d");
+        if verbose {
+            eprintln!("[cargo-miri rustc] writing dep-info to `{}`", dep_info_name.display());
```
Comment (RalfJung) [inline cargo-miri/bin.rs:617]:
```suggestion
            eprintln!("[cargo-miri rustc] writing stub dep-info to `{}`", dep_info_name.display());
```

File: cargo-miri/bin.rs [hunk 2]
```diff
@@ -609,6 +609,14 @@ fn phase_cargo_rustc(mut args: env::Args) {
...
+        }
+        File::create(dep_info_name).expect("failed to create fake .d file");
         let filename = out_filename("", "");
```
Comment (RalfJung) [inline cargo-miri/bin.rs:620]: Please add an empty line after the new code to logically separate it.

File: test-cargo-miri/run-test.py [hunk 3]
```diff
@@ -70,14 +73,20 @@ def test_no_rebuild(name, cmd):
...
-        env={
-            'MIRIFLAGS': "-Zmiri-disable-isolation",
-            'MIRITESTVAR': "wrongval", # make sure the build.rs value takes precedence
-        },
+        env=default_env,
+    )
+    # Special test: run it again *without* `-q` to make sure nothing is being rebuilt (Miri issue #1722)
+    test_no_rebuild("`cargo miri run` (no rebuild, no isolation)",
+        cargo_miri("run", quiet=False) + ["--", ""],
+        env=default_env,
```
Comment (RalfJung) [inline test-cargo-miri/run-test.py:89]: Why these other changes? Just moving the old command up should do it.

**Comment by RalfJung** [comment]
So this actually works :rofl: amazing.

File: test-cargo-miri/run-test.py (hunk 3 above)
Comment (hyd-dev) [reply test-cargo-miri/run-test.py]: > Just moving the old command up should do it.

Actually, no: https://github.com/rust-lang/miri/blob/4ecb5a99e09f9d90e33bc6ce667c592c173175a3/test-cargo-miri/build.rs#L15
But yes, `MIRIFLAGS` can be removed. That's indeed an oversight, sorry.

Comment (RalfJung) [reply test-cargo-miri/run-test.py]: Ah, good catch... I am not very happy about the code duplication here but whatever.

File: test-cargo-miri/run-test.py [hunk 4]
```diff
@@ -79,6 +82,11 @@ def test_cargo_miri_run():
             'MIRITESTVAR': "wrongval", # make sure the build.rs value takes precedence
         },
     )
+    # Special test: run it again *without* `-q` to make sure nothing is being rebuilt (Miri issue #1722)
+    test_no_rebuild("`cargo miri run` (no rebuild, no isolation)",
+        cargo_miri("run", quiet=False) + ["--", ""],
+        env={'MIRITESTVAR': "wrongval"},
```
Comment (RalfJung) [inline test-cargo-miri/run-test.py:88]:
```suggestion
        env={'MIRITESTVAR': "wrongval"}, # changing the env var causes a rebuild, so keep it set
```

File: test-cargo-miri/run-test.py [hunk 5]
```diff
@@ -79,6 +82,11 @@ def test_cargo_miri_run():
             'MIRITESTVAR': "wrongval", # make sure the build.rs value takes precedence
         },
     )
+    # Special test: run it again *without* `-q` to make sure nothing is being rebuilt (Miri issue #1722)
+    test_no_rebuild("`cargo miri run` (no rebuild, no isolation)",
```
Comment (RalfJung) [inline test-cargo-miri/run-test.py:86]:
```suggestion
    test_no_rebuild("`cargo miri run` (no rebuild)",
```

File: test-cargo-miri/run-test.py (hunk 3 above)
Comment (hyd-dev) [reply test-cargo-miri/run-test.py]: > I am not very happy about the code duplication here

Is this better?
```python3
    wrongval = "wrongval"
    test("`cargo miri run` (no isolation)",
        cargo_miri("run"),
        "run.default.stdout.ref", "run.default.stderr.ref",
        stdin=b'12\n21\n',
        env={
            'MIRIFLAGS': "-Zmiri-disable-isolation",
            'MIRITESTVAR': wrongval, # make sure the build.rs value takes precedence
        },
    )
    # Special test: run it again *without* `-q` to make sure nothing is being rebuilt (Miri issue #1722)
    test_no_rebuild("`cargo miri run` (no rebuild)",
        cargo_miri("run", quiet=False) + ["--", ""],
        env={'MIRITESTVAR': wrongval}, # changing the env var causes a rebuild (re-runs build.rs),
                                       # so keep it set
    )
```
I'm not familiar with Python so I don't know if I can deduplicate them further...

Comment (RalfJung) [reply test-cargo-miri/run-test.py]: I meant the duplication between `test_no_rebuild` and `test`... but it's not a big deal, we barely touch this code anyway.

**Comment by RalfJung** [comment]
Thanks a lot. :)
@bors r+

File: test-cargo-miri/run-test.py (hunk 3 above)
Comment (hyd-dev) [reply test-cargo-miri/run-test.py]: Oh... Sorry for misinterpreting the comment... 😰

Comment (RalfJung) [reply test-cargo-miri/run-test.py]: No worries, I was not very clear.

**Pull request merged**